cmd/sshpiperd/internal/plugin/grpc.go

```diff
@@ -332,6 +332,14 @@ func (g *GrpcPlugin) createUpstream(conn ssh.ConnMetadata, challengeCtx ssh.Chal
 		if err != nil {
 			return nil, err
 		}
+
+		if caPublicKey, _, _, _, err := ssh.ParseAuthorizedKey(a.GetCaPublicKey()); err == nil {
+			private, err = ssh.NewCertSigner(caPublicKey.(*ssh.Certificate), private)
```

Comment:
need to check if publickey is ssh.certificate?

Reply:
If `ssh.ParseAuthorizedKey` is successful, we known we will get back a `ssh.PublicKey` and that can be converted to a `ssh.Certificate`. Are you suggesting we add an extra layer of verification the it was able to convert the type of `ssh.PublicKey` to `ssh.Certificate`?

Reply:
seems not 100%
https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.12.0:ssh/keys.go;l=79

Reply:
Sorry, is this what you mean:

```
if a := upstream.GetPrivateKey(); a != nil {
	private, err := ssh.ParsePrivateKey(a.GetPrivateKey())
	if err != nil {
		return nil, err
	}

	if caPublicKey := a.GetCaPublicKey(); caPublicKey != nil {
		caPublicKey, _, _, _, err := ssh.ParseAuthorizedKey(caPublicKey)
		if err != nil {
			return nil, err
		}
		private, err = ssh.NewCertSigner(caPublicKey.(*ssh.Certificate), private)
		if err != nil {
			return nil, err
		}
	}

	config.Auth = append(config.Auth, ssh.PublicKeys(private))
	auth = append(auth, "privatekey")
}
```

If not can you give me a code example, I am not quite understanding what you mean.